README.md
# Simple Decimal
Java library with more convenient class alternative BigDecimal!

### Download
Maven:
```xml
<dependency>
  <groupId>xyz.ukrainskiys</groupId>
  <artifactId>simple-decimal</artifactId>
  <version>1.0</version>
</dependency>
```
Gradle:
```groovy
dependencies {
  implementation 'xyz.ukrainskiys:simple-decimal:1.0'
}
```

### Documentation
##### Create instance:

```java
import xyz.ukrainskiys.decimal.Decimal;

class Main {

  public static void main(String[] args) {
    var decimal = new Decimal(); // 0
    var decimal = new Decimal(BigDecimal.valueOf(666)); // 666
    var decimal = Decimal.of(666); // 666
    var decimal = Decimal.of(6.66); // 6.66
    var decimal = Decimal.ofKopecks(12345); // 123.45
  }
}
```

##### Constants:

```java
Decimals.ZERO;
Decimals.ONE;
Decimals.TEN; 
Decimals.HUNDRED;
Decimals.THOUSAND;
```

##### Comparison:
```java
Decimals.HUNDRED.more(Decimals.TEN); // true   
Decimals.HUNDRED.less(Decimals.TEN); // false   
Decimals.HUNDRED.more(100); // false
    
Decimal.of(10).max(Decimal.of(15)); // 15
    
Decimal.of(10).is(BigDecimal.TEN); // true   
Decimal.of(10.00).is(10); // true
```

##### Default operations:
```java
import xyz.ukrainskiys.decimal.constant.Decimals;

class Main {

  public static void main(String[] args) {
    final var decimal = Decimals.HUNDRED;
    
    decimal.add(Decimals.ONE); // 101
    decimal.subtract(1.5); // 99.5
    decimal.multiply(5); // 500
    decimal.safeDivide(Decimals.TEN); // 10
    decimal.safeDivide(33); // 3.0303030303
  }
}
```

##### Additional features:
```java
Decimal.of(200).ofPercent(50); // 100
    
Decimal.of(3.0303030303).safeSetScale(1); // 3.0   
Decimal.of(3.0303030303).setScaleWithKopecks(); // 3.03
Decimal.of(3.0303030303).setScale(2, RoundingMode.HALF_UP); // 3.03 (Standard BigDecimal method)
```

##### JPA:
For use the Decimal class with JPA, you must specify a converter.
```java
import javax.persistence.Column;
import javax.persistence.Convert;
import javax.persistence.Entity;
import javax.persistence.Id;
import xyz.ukrainskiys.decimal.Decimal;
import xyz.ukrainskiys.decimal.converter.DecimalConverter;

@Entity
public class TestElement {

  @Id
  @Column(name = "id", nullable = false)
  private Long id;

  @Convert(converter = DecimalConverter.class)
  private Decimal decimal;
}
```
In the table, this will be the numeric type, like BigDecimal.
![Example](img/decimal_table_example.png)


CHANGELOG.md
##1.1
* Delete implementation Decimal from BigDecimal;

##1.2
* Added a converter for the Decimal class for use with the JPA Repository;
* Added .toString() method fot the Decimal class;


src/main/java/xyz/ukrainskiys/decimal/Decimal.java
package xyz.ukrainskiys.decimal;

import java.io.Serializable;
import java.math.BigDecimal;
import java.math.RoundingMode;
import xyz.ukrainskiys.decimal.constant.Decimals;

public class Decimal extends Number implements Comparable<Decimal>, Serializable {

    private final BigDecimal bigDecimal;

    /**
     * <pre>{@code
     * new Decimal() == 0
     * new Decimal(BigDecimal.valueOf(666)) == 666
     * Decimal.of(666) == 666
     * Decimal.of(6.66) == 6.66
     * Decimal.ofKopecks(12345) == 123.45}</pre>
     */
    public Decimal(BigDecimal bigDecimal) {
        this.bigDecimal = bigDecimal;
    }

    public Decimal() {
        this.bigDecimal = BigDecimal.ZERO;
    }

    public static Decimal of(int value) {
        return new Decimal(BigDecimal.valueOf(value));
    }

    public static Decimal of(long value) {
        return new Decimal(BigDecimal.valueOf(value));
    }

    public static Decimal of(double value) {
        return new Decimal(BigDecimal.valueOf(value));
    }

    public static Decimal of(String value) {
        return new Decimal(new BigDecimal(value));
    }

    public static Decimal of(BigDecimal value) {
        return new Decimal(DecimalUtils.safeGet(value));
    }

    public static Decimal ofKopecks(long kopecks) {
        return new Decimal(BigDecimal.valueOf(kopecks)).safeDivide(100);
    }

    /**
     * <pre>{@code
     * Decimal.HUNDRED.more(Decimal.TEN) == true
     * Decimal.HUNDRED.less(Decimal.TEN) == false
     * Decimal.HUNDRED.more(100) == false}</pre>
     */
    public boolean more(Decimal other) {
        return bigDecimal.compareTo(other.bigDecimal) > 0;
    }

    public boolean more(BigDecimal other) {
        return bigDecimal.compareTo(DecimalUtils.safeGet(other)) > 0;
    }

    public boolean more(int value) {
        return bigDecimal.compareTo(BigDecimal.valueOf(value)) > 0;
    }

    public boolean more(long value) {
        return bigDecimal.compareTo(BigDecimal.valueOf(value)) > 0;
    }

    public boolean more(double value) {
        return bigDecimal.compareTo(BigDecimal.valueOf(value)) > 0;
    }

    public boolean more(String value) {
        return bigDecimal.compareTo(new BigDecimal(value)) > 0;
    }

    public boolean less(Decimal other) {
        return bigDecimal.compareTo(other.bigDecimal) < 0;
    }

    public boolean less(BigDecimal other) {
        return bigDecimal.compareTo(DecimalUtils.safeGet(other)) < 0;
    }

    public boolean less(int value) {
        return bigDecimal.compareTo(BigDecimal.valueOf(value)) < 0;
    }

    public boolean less(long value) {
        return bigDecimal.compareTo(BigDecimal.valueOf(value)) < 0;
    }

    public boolean less(double value) {
        return bigDecimal.compareTo(BigDecimal.valueOf(value)) < 0;
    }

    public boolean less(String value) {
        return bigDecimal.compareTo(new BigDecimal(value)) < 0;
    }

    public Decimal max(Decimal value) {
        return Decimal.of(bigDecimal.max(value.bigDecimal));
    }

    public Decimal min(Decimal value) {
        return Decimal.of(bigDecimal.min(value.bigDecimal));
    }


    public Decimal add(Decimal other) {
        return Decimal.of(bigDecimal.add(other.bigDecimal));
    }

    public Decimal add(BigDecimal other) {
        return Decimal.of(bigDecimal.add(DecimalUtils.safeGet(other)));
    }

    public Decimal add(int value) {
        return add(Decimal.of(BigDecimal.valueOf(value)));
    }

    public Decimal add(long value) {
        return add(Decimal.of(BigDecimal.valueOf(value)));
    }

    public Decimal add(double value) {
        return add(Decimal.of(BigDecimal.valueOf(value)));
    }

    public Decimal add(String value) {
        return add(Decimal.of(new BigDecimal(value)));
    }


    public Decimal subtract(Decimal other) {
        return new Decimal(bigDecimal.subtract(other.bigDecimal));
    }

    public Decimal subtract(BigDecimal other) {
        return new Decimal(bigDecimal.subtract(DecimalUtils.safeGet(other)));
    }

    public Decimal subtract(int value) {
        return subtract(Decimal.of(BigDecimal.valueOf(value)));
    }

    public Decimal subtract(long value) {
        return subtract(Decimal.of(BigDecimal.valueOf(value)));
    }

    public Decimal subtract(double value) {
        return subtract(Decimal.of(BigDecimal.valueOf(value)));
    }

    public Decimal subtract(String value) {
        return subtract(Decimal.of(new BigDecimal(value)));
    }


    public Decimal multiply(Decimal other) {
        return new Decimal(bigDecimal.multiply(other.bigDecimal));
    }

    public Decimal multiply(BigDecimal other) {
        return new Decimal(bigDecimal.multiply(DecimalUtils.safeGet(other)));
    }

    public Decimal multiply(int value) {
        return multiply(BigDecimal.valueOf(value));
    }

    public Decimal multiply(long value) {
        return multiply(BigDecimal.valueOf(value));
    }

    public Decimal multiply(double value) {
        return multiply(BigDecimal.valueOf(value));
    }

    public Decimal multiply(String value) {
        return multiply(new BigDecimal(value));
    }

    /**
     * <pre>{@code
     * Decimal.HUNDRED.safeDivide(10) == 10
     * Decimal.HUNDRED.safeDivide(33) == 3.0303030303}</pre>
     */
    public Decimal safeDivide(Decimal other) {
        final var result = bigDecimal.divide(other.bigDecimal, 11, RoundingMode.HALF_UP);
        try {
            final var afterDot =  result.toString().split("\\.")[1].toCharArray();
            int count = afterDot.length - 1;
            while (count >= 0) {
                if (afterDot[count] != '0') {
                    break;
                }
                count--;
            }
            return new Decimal(result.setScale(++count, RoundingMode.HALF_UP));
        } catch (ArrayIndexOutOfBoundsException e) {
            return Decimals.ZERO;
        }
    }

    public Decimal safeDivide(BigDecimal other) {
        return safeDivide(Decimal.of(other));
    }

    public Decimal safeDivide(int value) {
        return safeDivide(Decimal.of(value));
    }

    public Decimal safeDivide(long value) {
        return safeDivide(Decimal.of(value));
    }

    public Decimal safeDivide(double value) {
        return safeDivide(Decimal.of(value));
    }

    public Decimal safeDivide(String value) {
        return safeDivide(Decimal.of(value));
    }

    /**
     * <pre>{@code
     * Decimal.of(200).ofPercent(50) == 100}</pre>
     */
    public Decimal ofPercent(int percent) {
        return multiply(percent).safeDivide(Decimals.HUNDRED);
    }

    /**
     * <pre>{@code
     * Decimal.of(10).is(BigDecimal.TEN) == true
     * Decimal.of(10.00).is(10) == true}</pre>
     */
    public boolean is(Decimal other) {
        return bigDecimal.compareTo(other.bigDecimal) == 0;
    }

    public boolean is(BigDecimal other) {
        return bigDecimal.compareTo(DecimalUtils.safeGet(other)) == 0;
    }

    public boolean is(int value) {
        return intValue() == value;
    }

    public boolean is(double value) {
        return doubleValue() == value;
    }

    public boolean is(long value) {
        return longValue() == value;
    }

    /**
     * <pre>{@code
     * Decimal.of(3.0303030303).safeSetScale(1) == 3.0
     * Decimal.of(3.0303030303).setScaleWithKopecks() == 3.03}</pre>
     */
    public Decimal safeSetScale(int scale) {
        return new Decimal(bigDecimal.setScale(scale, RoundingMode.HALF_UP));
    }

    public Decimal setScaleWithKopecks() {
        return safeSetScale(2);
    }


    public BigDecimal getBigDecimal() {
        return bigDecimal;
    }

    @Override
    public int intValue() {
        return bigDecimal.intValue();
    }

    @Override
    public long longValue() {
        return bigDecimal.longValue();
    }

    @Override
    public float floatValue() {
        return bigDecimal.floatValue();
    }

    @Override
    public double doubleValue() {
        return bigDecimal.doubleValue();
    }

    @Override
    public int compareTo(Decimal o) {
        return bigDecimal.compareTo(o.bigDecimal);
    }

    @Override
    public String toString() {
        return bigDecimal.toString();
    }
}
